apollo-router/src/trace.rs

```diff
@@ -62,7 +62,16 @@ pub(crate) fn try_initialize_subscriber(
             let provider = builder.with_span_processor(batch).build();
 
             let tracer = provider.tracer("opentelemetry-jaeger", Some(env!("CARGO_PKG_VERSION")));
-            let _ = opentelemetry::global::set_tracer_provider(provider);
+
+            // The call to set_tracer_provider() manipulate a sync RwLock.
+            // Even though this code is sync, it is called from within an
+            // async context. If we don't do this in a separate thread,
+            // it will cause issues with the async runtime that prevents
+            // the router from working correctly.
+            let _ = std::thread::spawn(|| {
```

Comment:
joining on a thread from inside an async context might give the same result as waiting on a RwLock, since joining is blocking. But I don't see another way to do it (except maybe through channels with that thread)

Reply:
It's not ideal, but, for whatever reason, it's certainly better behaved with this fix in than otherwise.

Reply:
would tokio::spawn_blocking work better?

Reply:
I tried `spawn_blocking()` first and it also hangs.

Reply:
I thought the thread spawn looked cleaner. I need a way to wait on the spawn_blocking and the code looks a little bit messier (to my eyes).
e.g.:
```
            let jh = tokio::task::spawn_blocking(|| {
                opentelemetry::global::set_tracer_provider(provider);
            });
            futures::executor::block_on(jh).expect("block on set tracer");
```
I don't have strong feelings. I don't think either is more "correct" than the other.

Reply:
Ok. I used spawn_blocking(), since that's what it's designed for.
resolved in: bef9051cd1767f10423476cea1517dd68e5a7c09